### PR TITLE
[Merged by Bors] - 3315 Genesis address format - address size and type change

### LINKED ISSUE #3315: Genesis address format
See https://github.com/spacemeshos/pm/issues/140

- [x] choose and import a bech32 implementation (see [this list](https://github.com/spacemeshos/pm/issues/106#issuecomment-1171613630)) (chosen github.com/cosmos/btcutil)
- [x] use this to write a simple address parser (string -> `Address`) with error-handling code for wrong format, too long, bad HRP, etc. (this should probably be published as a separate library so it can be used elsewhere)
- [x] change `Address` type to `[24]byte` (i.e., change `AddressLength`)
- [x] update API code to receive address as string and parse it
  - [x] go-spacemesh (https://github.com/spacemeshos/go-spacemesh/pull/3336)
  - [x] API (https://github.com/spacemeshos/api/pull/173)
- [x] update addresses in genesis account config and tests
- [x] register HRPs (mainnet, testnet) in [SLIP-0713](https://github.com/satoshilabs/slips/blob/master/slip-0173.md)
  - [x] wait pr to be merged https://github.com/satoshilabs/slips/pull/1386

### PR DESCRIPTION
**Pull request by abergasov**
## Motivation

Closes #3315

## Changes
* address moved to separate pkg. probably in future will be as independent library
* address still array of bytes ([24]byte), but implement `bech32` algorithm. Additional first byte reserved for contain HumanReadablePart of address (now contain info about networkID - test|main net), so result is [25]byte
* all api endpoints will be changed to use string instead []byte (related pr https://github.com/spacemeshos/api/pull/173)

## Test Plan
unit tests

## TODO

- [ ] Explain motivation or link existing issue(s)
- [ ] Test changes and document test plan
- [ ] Update documentation as needed

## DevOps Notes

- [ ] This PR does not require configuration changes (e.g., environment variables, GitHub secrets, VM resources)
- [ ] This PR does not affect public APIs
- [x] This PR does not rely on a new version of external services (PoET, elasticsearch, etc.)
- [x] This PR does not make changes to log messages (which monitoring infrastructure may rely on)

### CODE REVIEW

**Comment by dshulyak** [comment]
could you please use aliases for such refactoring?

```go
package types

type Address = address.Address
```

they were created specifically to make such changes reviewable

**Review by dshulyak — CHANGES_REQUESTED** [review]
hrp part wont be stored on the mesh, even if it is 1 byte. it has other implications beside space

moreover:
- there will be no network ids. they will be replaced by hashes instead. so the idea to compress hrp wont work, but it is not even important
- there is no reason to hardcode every single known testnet name in advance. there are solutions that avoid that

File: common/types/address/address_test.go
Comment (dshulyak) [inline common/types/address/address_test.go]: rm

File: api/grpcserver/mesh_service.go
Comment (dshulyak) [inline api/grpcserver/mesh_service.go]: please stick to logs with fields

also:
- this is a warning, not the error
- it might be already logged by grpc middleware. please remove such logs if this is the case

File: common/types/address/address.go
Comment (dshulyak) [inline common/types/address/address.go]: FromString

File: common/types/address/config.go
Comment (dshulyak) [inline common/types/address/config.go]: i think only hrp should be required to configure. why do we need NetworkName?

File: cmd/node/node.go
Comment (dshulyak) [inline cmd/node/node.go:608]: clearly can't panic in API request, return error to notify node administrator that node is miconfigured

File: api/config/genesis.go
Comment (dshulyak) [inline api/config/genesis.go]: please remove this. either:

- use default hrp for this addresses
- check where this constants are used and remove them

File: api/grpcserver/globalstate_service.go
Comment (dshulyak) [inline api/grpcserver/globalstate_service.go]: it should be string with hrp

File: api/grpcserver/globalstate_service.go
Comment (dshulyak) [inline api/grpcserver/globalstate_service.go]: remove or downgrade and use fields

File: api/grpcserver/mesh_service.go
Comment (dshulyak) [inline api/grpcserver/mesh_service.go]: same

File: common/types/address/address.go
Comment (dshulyak) [inline common/types/address/address.go]: should be removed, this is replaced with bech32 string

File: common/types/address/address.go
Comment (dshulyak) [inline common/types/address/address.go]: this needs to be fixed, since String will include hrp

File: common/types/address/address.go
Comment (dshulyak) [inline common/types/address/address.go]: please double check with @lrettig if we need to leave 4 top bytes empty, he mentioned it somewhere in the github issues

**Review by dshulyak — CHANGES_REQUESTED** [review]
why cosmos/bech32 and not one from btcd repo? are they different?

File: common/types/address/address.go
Comment (lrettig) [reply common/types/address/address.go]: NewAddressFromString?

File: common/types/address/address.go
Comment (lrettig) [reply common/types/address/address.go]: correct, thanks @dshulyak

File: common/types/address/address.go
Comment (dshulyak) [reply common/types/address/address.go]: address is a part of the module, so what looks better when you use it?

address.NewAddressFromString or address.FromString

ps: i think the 1st one is incredibly ugly)

File: api/grpcserver/globalstate_service.go
Comment (abergasov) [reply api/grpcserver/globalstate_service.go]: it will be changed, when PR  https://github.com/spacemeshos/api/pull/173 will be merged
i can add cast to string, which will be excess after api method change signature, or just wait until pr merge, that i'll replace to `FromString`

File: cmd/node/node.go
Comment (abergasov) [reply cmd/node/node.go:608]: this is not an api request, it node starting. On next lines, panic can happen if address check failing
```
if coinbaseAddr.Big().Uint64() == 0 {
	app.log.Panic("invalid coinbase account")
}
```

File: common/types/address/address.go
Comment (abergasov) [reply common/types/address/address.go]: i removed this method, as it used for 1 plase in logs. Just change to `String`

File: common/types/address/config.go
Comment (abergasov) [reply common/types/address/config.go]: we do not need it. Probably we will later stream it via api for better ui usability?

File: cmd/node/node.go
Comment (dshulyak) [reply cmd/node/node.go:608]: sorry :) for some reason i thought that i saw same line in two places

**Comment by abergasov** [comment]
> why cosmos/bech32 and not one from btcd repo? are they different?

source code almost the same. small difference between const|wrappers, but in generally code the same. it can easy replaced by fixing `go.mod`

i chose cosmos cause it's approach - to get necessary functions and types, for crypto projects. in case btcd repo - we need to vendor other blockchein implementation, that's look little bit weird for me.

File: api/config/genesis.go
Comment (abergasov) [reply api/config/genesis.go]: in `address.BytesToAddress` will be used default hrp from `address.Config`

this consts are in `api/config/genesis.go`
they are used in `DefaultGenesisConfig` function 
```
type GenesisConfig struct {
	Accounts map[string]uint64 `mapstructure:"accounts"`
}

// Account1Address is the address from Account1Private.
const Account1Address = "0x0002abedccd3ffcbf46..."

// Account2Address is the address from Account2Address.
const Account2Address = "0xfeaf3d38b6ef43093..."

func DefaultGenesisConfig() *GenesisConfig {
	// NOTE(dshulyak) keys in default config are used in some tests
	g := GenesisConfig{}

	// we default to 10^5 SMH per account which is 10^17 smidge
	// each genesis account starts off with 10^17 smidge
	g.Accounts = map[string]uint64{
		Account1Address: 100000000000000000,
		Account2Address: 100000000000000000,
	}
	return &g
}
```

while able situation, when user can override with bech32 address like `sm1abc...`. 
So we should add limitation for config genesis address (like not use bech32 in address, but put `0x0123` string for genesis addr or other encoding style). Or just use such `if else`

Comment (dshulyak) [reply api/config/genesis.go]: sorry i don't follow. please replace 0x addresses with addresses prefixed using hrp and remove this if condition. spacecraft and system tests have to be adjusted accordingly.

File: common/types/address/address.go
Comment (abergasov) [reply common/types/address/address.go]: done, added bytes reservation + tests

File: common/types/address/config.go
Comment (dshulyak) [reply common/types/address/config.go]: i don't know why we would need to stream it. for now it just looks like totally unnecessary config option that will have be changed for every network. lets remove it

**Review by countvonzero — COMMENTED** [review]
i am generally surprised by the number of files changed for this PR.

what is the advantage of having a separate pkg for address?
in another words, what is the disadvantage of maintaining the flat structure inside common/types?

**Comment by abergasov** [comment]
> i am generally surprised by the number of files changed for this PR.
> 
> what is the advantage of having a separate pkg for address? in another words, what is the disadvantage of maintaining the flat structure inside common/types?

alias `type Address = address.Address` in common/type reduce total edited files =)

one of idea was to move address implementation to separate library, which will be included as dependency.
in this case  i moved to nested pkg. when we will publish as single library - just change imports

**Comment by lrettig** [comment]
> one of idea was to move address implementation to separate library, which will be included as dependency

This is not a terrible idea - do we think other projects will have this dependency? (And would they want to use just an address module without relying on all of go-spacemesh?) E.g., sm-repl, dash/explorer, etc.

**Comment by countvonzero** [comment]
> > one of idea was to move address implementation to separate library, which will be included as dependency
> 
> This is not a terrible idea - do we think other projects will have this dependency? (And would they want to use just an address module without relying on all of go-spacemesh?) E.g., sm-repl, dash/explorer, etc.

but why? what is the problem with having the impl inside common/types/address.go? the functionality you are providing in the new pkg look highly similar to the old ones. @abergasov

File: common/types/address/address_test.go
Comment (lrettig) [inline common/types/address/address_test.go]: please test too long as well as too short

also, bad bech32 formats (missing hrp, invalid characters, etc.)

File: common/types/address/config.go
Comment (lrettig) [inline common/types/address/config.go]: "smt" is too similar to "sm", this is too error-prone. per https://github.com/spacemeshos/pm/issues/106#issuecomment-1170509194, let's use "stest" for the testnet.

**Comment by abergasov** [comment]
> This is not a terrible idea - do we think other projects will have this dependency? (And would they want to use just an address module without relying on all of go-spacemesh?) E.g., sm-repl, dash/explorer, etc.

there is no strong arguments that we should move it to separate pkg, only this 2:
1. our possibly moving `address` in separate lib in future 
1. current situation in `common/types`, where we have 25 different files aka 25 different types (there are total 33 files with tests) and some of this files contains lot of code - this look like place where we can add more pkg separation and make it more strict

if you think that it better move back to `common/types` - ok, i'll do. arguments for separate pkg not very strong

**Review by lrettig — CHANGES_REQUESTED** [review]

**Comment by countvonzero** [comment]
> > This is not a terrible idea - do we think other projects will have this dependency? (And would they want to use just an address module without relying on all of go-spacemesh?) E.g., sm-repl, dash/explorer, etc.
> 
> there is no strong arguments that we should move it to separate pkg, only this 2:
> 
> 1. our possibly moving `address` in separate lib in future

i think this is a future thing and should be separated from this discussion.

> 2. current situation in `common/types`, where we have 25 different files aka 25 different types (there are total 33 files with tests) and some of this files contains lot of code - this look like place where we can add more pkg separation and make it more strict

right. but changing only 1 makes it even more confusing. i'd support if you make a sweeping change later to change all of them. but the code inside each file (tho large) are mostly very simple code... i personally don't think it's worth the effort now.
 
> if you think that it better move back to `common/types` - ok, i'll do. arguments for separate pkg not very strong

that will be my preference. and worry about separate library when the use case arise.

File: common/types/address/address_test.go
Comment (abergasov) [reply common/types/address/address_test.go]: done
but actually it will just bech32 lib tests) all this cases will generate err on decode and it immediately returns

**Comment by dshulyak** [comment]
bors try

File: common/types/address.go
Comment (dshulyak) [inline common/types/address.go]: > contains slice 8 bytes unsigned integers

what does it mean? where are the integers?

File: common/types/address.go
Comment (dshulyak) [inline common/types/address.go]: it needs to be more specific now, since the module was removed. maybe StringToAddress if there is already BytesToAddress

File: common/types/address.go
Comment (dshulyak) [inline common/types/address.go:142]: copy(addr[4:], publicKey[len(publicKey)-24:])

File: common/types/address.go
Comment (dshulyak) [inline common/types/address.go:77]: it needs a check that first 4 bytes are empty

File: config/presets/testnet.go
Comment (dshulyak) [inline config/presets/testnet.go]: this are the same addresses just re-encoded or from different private keys?

File: common/types/address.go
Comment (abergasov) [reply common/types/address.go]: by default slice is uint8, value can be stored from 0-256 (base 2^8=256)
for bech32 need convert to 5-bit integers (2^5=32). it still uint8, but value up to 32
```
bech32.ConvertBits(addr[:], 8, 5, true) -> before encode to `sm1abc`
bech32.ConvertBits(addr[:], 5, 8, true) -> after encode from `sm1abc`
```

after this conversion slice length make bigger. so we keep as usually, uint8

File: common/types/address.go
Comment (abergasov) [reply common/types/address.go]: `BytesToAddress` will be dropped, when we merge api changes
I'll rename to `StringToAddress`

File: common/types/address.go
Comment (dshulyak) [reply common/types/address.go]: please remove this comment. it doesn't store uint64 (8 bytes unsigned inteegers) in any shape. what you provided above is a specific encoding of the bech32 format

File: config/presets/testnet.go
Comment (abergasov) [reply config/presets/testnet.go]: yes, i make ` util.FromHex(str)` and create address from it
as we keep algorithm how convert bytes to address (get last 20 bytes from byte slice)  and we fill first 4 bytes with zero - we can just regenerate this addresses

**Comment by abergasov** [comment]
bors try

**Comment by dshulyak** [comment]
what about system tests?

**Comment by abergasov** [comment]
> what about system tests?

it fails in same place `cluster.Reuse` with ctx timeout. Probably node fails to start in case invalid genessis|coinbase accounts

**Comment by abergasov** [comment]
bors try

**Comment by abergasov** [comment]
bors try

**Comment by abergasov** [comment]
bors try

**Review by dshulyak — APPROVED** [review]

**Comment by abergasov** [comment]
bors merge

**Comment by abergasov** [comment]
bors merge

File: api/grpcserver/http_server.go
Comment (lrettig) [inline api/grpcserver/http_server.go]: Why rename this?

File: common/types/address.go
Comment (lrettig) [inline common/types/address.go]:
```suggestion
	ErrDecodeBech32 = errors.New("error decoding bech32")
```

File: common/types/address.go
Comment (lrettig) [inline common/types/address.go]:
```suggestion
		return addr, fmt.Errorf("error converting bech32 bits: %w", err)
```

File: common/types/address.go
Comment (lrettig) [inline common/types/address.go]:
```suggestion
		log.Panic("error encoding to bech32: ", err.Error())
```

File: common/types/address.go
Comment (lrettig) [inline common/types/address.go]:
```suggestion
		log.Panic("error converting bech32 bits: ", err.Error())
```

File: common/types/address_test.go
Comment (lrettig) [inline common/types/address_test.go]:
```suggestion
			name: "correct address, but no reserved space",
```

File: common/types/address_test.go
Comment (lrettig) [inline common/types/address_test.go]:
```suggestion
			name: "incorrect charset", // chars `iob` not supported by bech32.
```

File: common/types/address_test.go
Comment (lrettig) [inline common/types/address_test.go]:
```suggestion
			name: "too long",
```

File: common/types/address_test.go
Comment (lrettig) [inline common/types/address_test.go]:
```suggestion
			name: "too short",
```

File: common/types/address_test.go
Comment (lrettig) [inline common/types/address_test.go]:
```suggestion
			name: "wrong network (hrp)",
```

**Review by lrettig — CHANGES_REQUESTED** [review]

**Comment by abergasov** [comment]
bors try

**Review by lrettig — APPROVED** [review]

**Comment by dshulyak** [comment]
i thought that it was merged. is there any reason why it is still open?

**Comment by abergasov** [comment]
bors try

**Comment by abergasov** [comment]
bors merge

**Pull request closed without merging**
